File: backend/src/taxonomy_builder/schemas/feedback.py
"""Pydantic schemas for Feedback."""

from datetime import datetime
from typing import Self
from uuid import UUID

from pydantic import BaseModel, Field, field_validator, model_validator

from taxonomy_builder.models.feedback import (
    FEEDBACK_TYPES_BY_ENTITY,
    EntityType,
    FeedbackStatus,
)


class FeedbackCreate(BaseModel):
    """Schema for creating feedback."""

    snapshot_version: str
    entity_type: EntityType
    entity_id: str
    entity_label: str = Field(max_length=500)
    feedback_type: str = Field(max_length=100)
    content: str = Field(min_length=1, max_length=10000)

    @field_validator("content")
    @classmethod
    def strip_and_validate_content(cls, v: str) -> str:
        stripped = v.strip()
        if not stripped:
            raise ValueError("Content cannot be empty or whitespace only")
        return stripped

    @model_validator(mode="after")
    def validate_feedback_type_for_entity(self) -> Self:
        valid_types = FEEDBACK_TYPES_BY_ENTITY.get(self.entity_type)
        if valid_types and self.feedback_type not in valid_types:
            raise ValueError(
                f"feedback_type '{self.feedback_type}' is not valid "
                f"for entity_type '{self.entity_type.value}'"
            )
        return self


class FeedbackResponse(BaseModel):
    """Nested response info shown to readers (no manager identity)."""

    content: str
    created_at: datetime


class FeedbackRead(BaseModel):
    """Schema for reading feedback (reader-facing)."""

    id: UUID
    project_id: UUID
    snapshot_version: str
    entity_type: str
    entity_id: str
    entity_label: str
    feedback_type: str
    content: str
    status: FeedbackStatus
    response: FeedbackResponse | None = None
    created_at: datetime
    can_delete: bool = False


class FeedbackManagerRead(FeedbackRead):
    """Schema for reading feedback (manager-facing, includes author info)."""

    author_name: str
    responded_by_name: str | None = None


class RespondRequest(BaseModel):
    """Body for POST /respond."""

    content: str = Field(min_length=1, max_length=10000)

    @field_validator("content")
    @classmethod
    def strip_and_validate(cls, v: str) -> str:
        stripped = v.strip()
        if not stripped:
            raise ValueError("Response content cannot be empty or whitespace only")
        return stripped


class TriageRequest(BaseModel):
    """Body for POST /resolve and POST /decline."""

    content: str | None = Field(default=None, max_length=10000)

File: frontend/src/api/feedback.ts
import { api } from "./client";
import type { FeedbackManagerRead } from "../types/models";

export interface FeedbackListParams {
  status?: string;
  entity_type?: string;
  feedback_type?: string;
  q?: string;
}

export const feedbackManagerApi = {
  listAll: (projectId: string, params?: FeedbackListParams) => {
    const query = new URLSearchParams();
    if (params?.status) query.set("status", params.status);
    if (params?.entity_type) query.set("entity_type", params.entity_type);
    if (params?.feedback_type) query.set("feedback_type", params.feedback_type);
    if (params?.q) query.set("q", params.q);
    const qs = query.toString();
    return api.get<FeedbackManagerRead[]>(
      `/feedback/${projectId}/all${qs ? `?${qs}` : ""}`,
    );
  },

  respond: (feedbackId: string, content: string) =>
    api.post<FeedbackManagerRead>(`/feedback/${feedbackId}/respond`, {
      content,
    }),

  resolve: (feedbackId: string, content?: string) =>
    api.post<FeedbackManagerRead>(`/feedback/${feedbackId}/resolve`, {
      content: content || null,
    }),

  decline: (feedbackId: string, content?: string) =>
    api.post<FeedbackManagerRead>(`/feedback/${feedbackId}/decline`, {
      content: content || null,
    }),

  counts: (projectIds: string[]) => {
    const query = projectIds.map((id) => `project_ids=${id}`).join("&");
    return api.get<Record<string, number>>(`/feedback/counts?${query}`);
  },
};
